disco,net/tstun,wgengine/magicsock: probe and record peer MTU

valscale opened the pull request:
disco,net/tstun,wgengine/magicsock: probe and record peer MTU

Automatically probe and record the path MTU to a peer when peer MTU is enabled, 
but do not use the MTU information for anything other than metrics yet.

Updates #311

icio reviewed (commented):
init stuff looks good to me ☺️

icio commented on net/tstun/mtu.go:
```diff
@@ -79,14 +79,34 @@ const (
+// WireMTUsToProbe is a list of the on-the-wire MTUs we want to
+// probe.
+var WireMTUsToProbe = []WireMTU{
+	WireMTU(WireToTUNMTU(WireMTU(safeTUNMTU))), // Tailscale over Tailscale :)
```
If `safeTUNMTU` is already a TUNMTU, should we be explicitly type-casting it to `WireMTU` just to convert it back via `WireToTUNMTU`? Is safeTUNMTU actually a TUNMTU if it's not already accounting for wgHeaderLen?

valscale replied:
You're right, this is overly complicated. I just have to cast safeTUNMTU to a WireMTU and I'm done.

valscale commented on net/tstun/mtu.go:
```diff
@@ -79,14 +79,34 @@ const (
+func findMaxProbedWireMTU() (maxWireMTU WireMTU) {
+	for _, m := range WireMTUsToProbe {
+		// See maxTUNMTU comment for why this may be very small
+		if m > WireMTU(maxTUNMTU) {
```
Good question. The safeTUNMTU is definitely smaller since that's what we have been using for all these years.

valscale commented on net/tstun/mtu.go:
```diff
@@ -79,14 +79,34 @@ const (
+		}
+		maxWireMTU = max(m, maxWireMTU)
 	}
+	return
```
Ooh, can I ask about go coding style here? This is a "naked" return so it has to go there - but I could also put in the name of the return value so it's obvious that it is necessary. Is there a convention?

valscale replied:
I rewrote the commit to not need this line at all, but I'm still curious about the preferred idiom if you happen to know it. :)

valscale commented on wgengine/magicsock/endpoint.go:
```diff
@@ -35,6 +36,23 @@ import (
+	v4 := netip.MustParseAddr("1.1.1.1")
+	v6 := netip.MustParseAddr("2001::3456")
+	for _, m := range tstun.WireMTUsToProbe {
+		// XXX one init() relying on another init()
```
Thank you! Fixed.

valscale commented on wgengine/magicsock/endpoint.go:
```diff
@@ -35,6 +36,23 @@ import (
+	v6 := netip.MustParseAddr("2001::3456")
+	for _, m := range tstun.WireMTUsToProbe {
+		// XXX one init() relying on another init()
+		if m > tstun.WireMTU(tstun.MaxProbedWireMTU()) {
```
Thanks, fixed!

valscale commented on wgengine/magicsock/endpoint.go:
```diff
@@ -35,6 +36,23 @@ import (
+			continue
+		}
+		fmt.Printf("mtu %v v4 %v v6 %v\n", m, pktLenToPingSize(m, v4), pktLenToPingSize(m, v6))
+		mtuProbePingSizesV4 = append(mtuProbePingSizesV4, pktLenToPingSize(m, v4))
```
Thanks, they both seemed bleah to me so I appreciate the vote in favor of the bool. Fixed.

valscale commented on wgengine/magicsock/magicsock.go:
```diff
@@ -2780,6 +2781,11 @@ var (
+	metricMaxPeerMTU = clientmetric.NewGauge("magicsock_max_peer_mtu")
+	// metricPeerMTUsProbed collects the number of times we received an peer MTU
+	// probe response for a given MTU size.
+	metricPeerMTUsProbed = &metrics.LabelMap{Label: "MTU"}
```
Copied, thanks to you and @andrew-d !

valscale commented:
Update: I removed the entire concept of max probed MTU from this PR since that won't be necessary until the PTB generation PR. I removed any modification of the MTUs to probe - it's such a tiny optimization and we can add it later if necessary. And I split the current usage of default TUN MTU into "what to set the tailscale TUN to" and "what is the safe value to assume that the wire MTU is if we have no information about the path."

valscale commented:
Quick update to use @raggi's [nice new syncs.LoadOrInit()](https://github.com/tailscale/tailscale/pull/9689).
